pkg/sources/sources.go

```diff
@@ -0,0 +1,249 @@
+}
+
+type registrationHandler func(resource string) cache.SharedIndexInformer
+
+type eventHandler func(ctx context.Context, resource string, sources []string, updateDiffs []string) func(obj interface{})
+
+func NewRouter(mapper meta.RESTMapper, dynamicCli dynamic.Interface, log logrus.FieldLogger) *Router {
+	return &Router{
+		log:           log,
+		mapper:        mapper,
+		dynamicCli:    dynamicCli,
+		table:         make(map[string][]Entry),
+		bindings:      make(map[string]struct{}),
+		registrations: make(map[string]Registration),
+	}
+}
+
+func (r *Router) AddAnyBindingsByName(c config.IdentifiableMap[config.ChannelBindingsByName]) {
+	for _, byName := range c {
+		r.AddAnyBindings(byName.Bindings)
+	}
+}
+
+func (r *Router) AddAnyBindingsById(c config.IdentifiableMap[config.ChannelBindingsByID]) {
+	for _, byID := range c {
+		r.AddAnyBindings(byID.Bindings)
+	}
+}
+
+func (r *Router) AddAnyBindings(b config.BotBindings) {
+	for _, source := range b.Sources {
+		r.bindings[source] = struct{}{}
+	}
+}
+
+func (r *Router) GetBoundSources(sources config.IndexableMap[config.Sources]) config.IndexableMap[config.Sources] {
+	out := make(config.IndexableMap[config.Sources])
+	for name, t := range sources {
+		if _, ok := r.bindings[name]; ok {
+			out[name] = t
+		}
+	}
+	return out
+}
+
+func (r *Router) BuildTable(cfg *config.Config) *Router {
+	sources := r.GetBoundSources(cfg.Sources)
+	mergedEvents := mergeResourceEvents(sources)
+
+	for resource, resourceEvents := range mergedEvents {
+		eventRoutes := mergeEventRoutes(resource, sources)
+		r.buildTable(resource, resourceEvents, eventRoutes)
+	}
+	r.log.Debugf("sources routing table: %+v", r.table)
+	return r
+}
+
+func mergeResourceEvents(sources config.IndexableMap[config.Sources]) mergedEvents {
+	out := map[string]map[config.EventType]struct{}{}
+	for _, srcGroupCfg := range sources {
+		for _, resource := range srcGroupCfg.Kubernetes.Resources {
+			if _, ok := out[resource.Name]; !ok {
+				out[resource.Name] = make(map[config.EventType]struct{})
+			}
+			for _, e := range flattenEvents(resource.Events) {
+				out[resource.Name][e] = struct{}{}
+			}
+		}
+	}
+	return out
+}
+
+func mergeEventRoutes(resource string, sources config.IndexableMap[config.Sources]) map[config.EventType][]Route {
+	out := make(map[config.EventType][]Route)
+	for srcGroupName, srcGroupCfg := range sources {
+		for _, r := range srcGroupCfg.Kubernetes.Resources {
+			for _, e := range flattenEvents(r.Events) {
+				switch {
+				case resource == r.Name && e == config.UpdateEvent:
+					out[e] = append(out[e], Route{
+						source:     srcGroupName,
+						namespaces: r.Namespaces,
+						updateSetting: config.UpdateSetting{
+							Fields:      r.UpdateSetting.Fields,
+							IncludeDiff: r.UpdateSetting.IncludeDiff,
+						}})
+				case resource == r.Name && e != config.UpdateEvent:
+					out[e] = append(out[e], Route{source: srcGroupName, namespaces: r.Namespaces})
```

Comment:
We can simplify this:


```suggestion
				if resource != r.Name {
					continue
				}

				route := Route{source: srcGroupName, namespaces: r.Namespaces}

				if e == config.UpdateEvent {
					route.updateSetting = config.UpdateSetting{
						Fields:      r.UpdateSetting.Fields,
						IncludeDiff: r.UpdateSetting.IncludeDiff,
					}
				}
				
				out[e] = append(out[e], route)
```